### PR TITLE
[5.5] Maintain queue and connection on chained jobs

### PR DESCRIPTION
**Pull request by themsaid**
While doing:

```php
JobA::withChain([
       ( new JobB() ),
       ( new JobB() )
])->dispatch()
  ->onConnection('custom-connection')
  ->onQueue('custom-queue');
```

The selected connection and queue won't be used for chained jobs, only JobA. This PR fixes that.

If you explicitly define a queue/connection on a job, it'll be used instead of the main one.

```php
JobA::withChain([
       (new JobB())->onQueue('some_queue'),
       new JobB()
])->dispatch()
  ->onConnection('custom-connection')
  ->onQueue('custom-queue');
```

Here JobB will run on `some_queue` while the rest of the jobs will run on `custom-queue`

### CODE REVIEW

**Comment by ConnorVG** [comment]
Would it be possible to have something like:
```php
JobA::withChain([
       new JobB($a, $b, $c),
       (new JobB($x, $y, $z))->onConnection('foo-connection'),
       new JobC($a, $b, $c),
       new JobD
])->dispatch($param1, $param2, $param3)->onConnection('bar-connection')->onQueue('baz-queue');
```
?

**Comment by Artistan** [comment]
maybe a check on next to see if it isset already or not...

```php
         if (! empty($this->chained)) {
              dispatch(tap(unserialize(array_shift($this->chained)), function ($next) {
                  $next->chained = $this->chained;
                  if(is_null($next->connection)){
                      $next->onConnection($this->connection);
                  }
                  if(is_null($next->queue)){
                      $next->onQueue($this->queue);
                  }
              }));
          }
```

**Comment by themsaid** [comment]
Updated the PR with a check to make sure if a connection/queue exists we won't override it.

**Comment by Artistan** [comment]
This may need to be a 5.6 thing since it actually changes how it currently works?
Some people may "expect" it to go to the default queue.

File: src/Illuminate/Bus/Queueable.php [hunk 1]
```diff
@@ -96,6 +96,8 @@ public function dispatchNextJobInChain()
         if (! empty($this->chained)) {
             dispatch(tap(unserialize(array_shift($this->chained)), function ($next) {
```
Comment (Artistan) [inline src/Illuminate/Bus/Queueable.php]: Can you do ` function (Queueable $next) {` ?
This allows for better code inspection 👍

**Comment by themsaid** [comment]
@Artistan it's tricky yes, if you think of it in a way you might see it as breaking, in another way it looks like the expected behaviour, so leaving this open to see what Taylor and other contributors think.

**Comment by taylorotwell** [comment]
Hmm, hard to say. If you explicitly set the connection and queue on each chained job does it work?

**Comment by themsaid** [comment]
@taylorotwell 

```php
JobA::withChain([
    (new JobB())->onQueue('second'),
    (new JobC()),
])->dispatch()->onQueue('first');
```

Currently:

- JobA will run on `first` queue
- JobB on `second` queue
- JobC on `default` queue

With this PR:

- JobA will run on `first` queue
- JobB on `second` queue
- JobC on `second` queue

And for this case:

```php
JobA::withChain([
    (new JobB()),
    (new JobC()),
])->dispatch()->onQueue('first');
```

All jobs will run on the `first` queue, while currently JobA will run on the first and JobB and JobC will run on `default`.

**Comment by taylorotwell** [comment]
I think I'm OK with existing behavior for now. If you want to specify queue the chained jobs should run on you should do that when creating them, etc. I would be OK with this proposal if it could be done with new methods like "allOnQueue" or "allOnConnection" etc.

**Pull request closed without merging**